docs/source/examples/performance-training.rst

```diff
@@ -0,0 +1,408 @@
+
+In this example, we fine-tune a BERT model on a question-answering task with HuggingFace.
+
+This example:
+
+- has SkyPilot find a V100 instance on any cloud,
+- uses spot instances to save cost, and
+- uses checkpointing to recover preempted jobs quickly.
+
+.. code-block:: yaml
+
+  # bert_qa.yaml
+  name: bert-qa
+
+  resources:
+    accelerators: V100:1
+    use_spot: true  # Use spot instances to save cost.
+    disk_tier: best # using highest performance disk tier
+
+  file_mounts:
+    /checkpoint:
+      name: # NOTE: Fill in your bucket name
+      mode: MOUNT_CACHED
+
+  envs:
+    # Fill in your wandb key: copy from https://wandb.ai/authorize
+    # Alternatively, you can use `--env WANDB_API_KEY=$WANDB_API_KEY`
+    # to pass the key in the command line, during `sky jobs launch`.
+    WANDB_API_KEY:
+
+  # Assume your working directory is under `~/transformers`.
+  workdir: ~/transformers
+
+  setup: |
+    pip install -e .
+    cd examples/pytorch/question-answering/
+    pip install -r requirements.txt torch==1.12.1+cu113 --extra-index-url https://download.pytorch.org/whl/cu113
+    pip install wandb
+
+  run: |
+    cd examples/pytorch/question-answering/
+    python run_qa.py \
+      --model_name_or_path bert-base-uncased \
+      --dataset_name squad \
+      --do_train \
+      --do_eval \
+      --per_device_train_batch_size 12 \
+      --learning_rate 3e-5 \
+      --num_train_epochs 50 \
+      --max_seq_length 384 \
+      --doc_stride 128 \
+      --report_to wandb \
+      --output_dir /checkpoint/bert_qa/ \
+      --run_name $SKYPILOT_TASK_ID \
+      --save_total_limit 10 \
+      --save_steps 1000
+
+The highlighted lines add a bucket for checkpoints.
+As HuggingFace has built-in support for periodic checkpointing, we just need to pass the highlighted arguments to save checkpoints to the bucket.
+(See more on `Huggingface API <https://huggingface.co/docs/transformers/main_classes/trainer#transformers.TrainingArguments.save_steps>`__).
+To see another example of periodic checkpointing with PyTorch, check out `our ResNet example <https://github.com/skypilot-org/skypilot/tree/master/examples/spot/resnet_ddp>`__.
+
+We also set :code:`--run_name` to :code:`$SKYPILOT_TASK_ID` so that the logs for all recoveries of the same job will be saved
+to the same run in Weights & Biases.
+
+.. note::
+  The environment variable :code:`$SKYPILOT_TASK_ID` (example: "sky-managed-2022-10-06-05-17-09-750781_bert-qa_8-0") can be used to identify the same job, i.e., it is kept identical across all
+  recoveries of the job.
+  It can be accessed in the task's :code:`run` commands or directly in the program itself (e.g., access
+  via :code:`os.environ` and pass to Weights & Biases for tracking purposes in your training script). It is made available to
+  the task whenever it is invoked. See more about :ref:`environment variables provided by SkyPilot <sky-env-vars>`.
```

Comment:
Maybe this should be mentioned outside of just the BERT example, like farther up on this page or on the checkpointing section in the managed jobs page.